From 4fa969203355ebb9de9f9868619b3b4c23d7da7c Mon Sep 17 00:00:00 2001
From: tikanpet <tikanderp@gmail.com>
Date: Thu, 5 Mar 2026 15:20:06 +0200
Subject: [PATCH] LUKS2-encryption for EXTx-partition. So the encrypted
 partition is included directly in WIC-image when building the image. New
 function 'encrypt_rootfs_ext()' is added in
 'poky/scripts/lib/wic/partition.py', which encrypts the root-filesystem using
 Cryptsetup. encrypt_rootfs_ext(), and changes for growing the partition for
 LUKS2-header are patched with 'wic-tools-encrypt.bb'. Encrypting is done
 afterwards. So filesystem-data is already available in the partition:

  "reencrypt --disable-locks --reduce-device-size 16M
  --force-offline-reencrypt --keyfile luks.key.."

This way the usage of 'sudo' is avoided, and Yocto remains sudoless.

So create the encryption key (keyfile) using Openssl-tool, and store as
  'poky/build/conf/luks.key'.
Also passphrase is added into the encrypted partition, which gives
a convenient way for the user for opening the encrypted partition in
target device (in Initramfs). So create the passphrase key, and store as
  'poky/build/conf/passphrase_luks.key'.

Also next new flag has to be added in 'poky/build/local.conf' for enabling the encryption:
   WICVARS:append = " LUKS2_ENCRYPT"
   LUKS2_ENCRYPT = "1"

Upstream-Status: Pending
Signed-off-by: tikanpet <tikanderp@gmail.com>
---
 scripts/lib/wic/partition.py | 90 ++++++++++++++++++++++++++++++++++++
 1 file changed, 90 insertions(+)

diff --git a/scripts/lib/wic/partition.py b/scripts/lib/wic/partition.py
index bf2c34d594..98f7a5e8a1 100644
--- a/scripts/lib/wic/partition.py
+++ b/scripts/lib/wic/partition.py
@@ -20,6 +20,9 @@ from wic.pluginbase import PluginMgr
 
 logger = logging.getLogger('wic')
 
+# Size of encryption header (16 MB in 1kB-blocks)
+LUKS2_HEADER_SIZE = 16000
+
 class Partition():
 
     def __init__(self, args, lineno):
@@ -279,6 +282,10 @@ class Partition():
 
         rootfs_size = self.get_rootfs_size(actual_rootfs_size)
 
+        # reserve some more space for encryption
+        if get_bitbake_var("LUKS2_ENCRYPT") == "1":
+            rootfs_size = rootfs_size + (3 * LUKS2_HEADER_SIZE)
+
         with open(rootfs, 'w') as sparse:
             os.ftruncate(sparse.fileno(), rootfs_size * 1024)
 
@@ -333,6 +340,89 @@ class Partition():
 
         self.check_for_Y2038_problem(rootfs, native_sysroot)
 
+        self.encrypt_rootfs_ext(rootfs, rootfs_size, native_sysroot, pseudo)
+
+    def encrypt_rootfs_ext(self, rootfs, rootfs_size, native_sysroot, pseudo):
+        """
+        LUKS2-encryption for ext-partition (root-fs).
+        """
+        if get_bitbake_var("LUKS2_ENCRYPT") != "1":
+            # encryption disabled
+            return
+
+        logger.info("Encrypt the partition (LUKS2) \n")
+        logger.info("Label = %s, disk = %s, filesystem type = %s) \n"
+                     %(self.label, self.disk, self.fstype))
+
+        # Encrypt the file system with LUKS (using 'CryptSetup'-tool).
+        # LUKS-header (LUKS2-version by default) requires some space in the beginning
+        # of the partition. So partition-size will first be reduced.
+        # Sparse file ('rootfs') will then have an empty space in the end of the file.
+        # Payload-data (root-file system) will next be shifted by 'Cryptsetup' for filling
+        # the empty space, and leaving the beginning of the partition for LUKS-header.
+        # Then encryption will be done by:
+        #'   --reencrypt...--reduce-device-size 16M'
+
+        # In this way the root-fs (payload data already included) can be encrypted,
+        # and LUKS2-header be inserted (by shifting existing payload data 16MB further).
+        # Notice that instead of using 'reencrypt', next procedure could have been tried:
+        #   'cryptsetup luksFormat...'
+        #   'sudo cryptsetup open /dev/mapper/decrypt'
+        #   'mkfs.ext4 .../dev/mapper/decrypt...-d rootfs_dir...'
+        #   'sudo mount /dev/mapper/decrypt /mnt'
+        #   'sudo cryptsetup close /dev/mapper/decrypt'
+        # But as a consequence, root-owner accesses were replaced by user-owner accesses
+        # for all root-files and -directories...
+        # Also there are a lot of sudo-accesses required. Yocto-environment is sudoless.
+        # 'Pseudo' commands (previously 'fakeroot') are used when acting as a root-user.
+        # But, 'pseudo'-rooting doesn't work, when trying to access
+        # e.g. block-devices (/dev/...).
+        # Well, 'reencrypt' would need root-access as well,
+        # although just 'rootfs'-file is accessed instead of '/dev/mapper/decrypt...'.
+        # The reason seems to be the usage of '/run/cryptsetup' directory for storing
+        # the runtime data during the encyption...
+        # Using 'PSEUDO' (to avoid 'sudoing') didn't work out either...
+
+        # Luckily '--reencrypt' has 'disable-locks'-option for allowing operations
+        # without using the '/run' directory for lock protection.
+
+        # Two encryption keys added. One key in such would be sufficient amount...
+        encrypt_keyfile = "%s/conf/luks.key" % (os.environ.get("BUILDDIR"))
+        if not os.path.isfile(encrypt_keyfile):
+            raise WicError("Encryption key for partition LUKS-encryption not found, exiting.")
+
+        encrypt_passphrase_keyfile = "%s/conf/passphrase_luks.key" % (os.environ.get("BUILDDIR"))
+        if not os.path.isfile(encrypt_passphrase_keyfile):
+            raise WicError("Secondary Encryption key (passphrase) not found, exiting.")
+
+        # LUKS2 header ~16 MB -> lets double that for being sure, we have enough space...
+        reduced_block_size = rootfs_size - (2 * LUKS2_HEADER_SIZE)
+
+        crypt_cmd = "resize2fs -p %s %s" %(rootfs, reduced_block_size)
+        exec_native_cmd(crypt_cmd, native_sysroot, None)
+
+        # Run the encryption process
+        crypt_cmd = "cryptsetup -v -q reencrypt --encrypt --key-file \
+                     %s --progress-frequency 1 --reduce-device-size 16M \
+                     --disable-locks --force-offline-reencrypt %s" % (encrypt_keyfile, rootfs)
+        exec_native_cmd(crypt_cmd, native_sysroot, None)
+
+        # add secondary LUKS-key (passphrase), which is default way for opening
+        # the LUKS-encryption in the target device (in InitRamFS)
+        crypt_cmd = "cryptsetup luksAddKey -v -q %s --new-keyfile %s --key-file %s" \
+                      % (rootfs, encrypt_passphrase_keyfile, encrypt_keyfile)
+        ret, out = exec_native_cmd(crypt_cmd, native_sysroot, None)
+
+        if ret != 0:
+            raise WicError("Encryption failed")
+
+        logger.info("Encryption succesfull")
+
+        # Filesystem is now encrypted, and two security keys are added.
+        # Resizing for using the whole partition (resize2fs /dev/mapper/decrypt)
+        # is still required, but to avoid root-accesses here (opening the LUKS-container),
+        # it's better to execute in Initramfs during booting of the target device.
+
     def get_hash_seed_ext4(self, extraopts, pseudo):
         if os.getenv('SOURCE_DATE_EPOCH'):
             sde_time = int(os.getenv('SOURCE_DATE_EPOCH'))
-- 
2.43.0

